### PR TITLE
ARROW-4520: [C++] Regression in unreachable DCHECK custom error messages

### PR DESCRIPTION
**Pull request by bkietz**
I'm compiling with clang-7 on Ubuntu 18.04 using conda to manage dependencies

cmake command:

```
cmake -GNinja \
 -DCMAKE_BUILD_TYPE=Release \
 -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX \
 -DARROW_PARQUET=on \
 -DARROW_PLASMA=on \
 -DBUILD_WARNING_LEVEL=CHECKIN \
 -DCMAKE_EXPORT_COMPILE_COMMANDS=on \
 -DARROW_BUILD_BENCHMARKS=on \
 ../..
```

On master this produces errors like:

```
../../src/arrow/type.cc:175:7: error: code will never be executed [-Werror,-Wunreachable-code]
      << "Must be microseconds or nanoseconds";
      ^~
```

### CODE REVIEW

**Comment by wesm** [comment]
On what system configuration to you get this issue? I haven't seen the unreachable error here with either gcc 4.8, 7.3, or 8.2

**Comment by wesm** [comment]
I see, with clang-7. Can you add information about this to the PR description?

**Comment by wesm** [comment]
Hm. How important is it for this to go into the 0.13 release?

**Comment by fsaintjacques** [comment]
Doesn't that revert the previous non-ignore bug?

**Comment by bkietz** [comment]
@wesm I'd say it's important that we can build in release config with clang-7
@fsaintjacques This doesn't evaluate expressions in release: https://github.com/apache/arrow/pull/4027/files#diff-04220e5f8662d031cedc678c7d48c00cR91

File: cpp/src/arrow/util/logging.h [hunk 1]
```diff
@@ -81,23 +81,36 @@ enum class ArrowLogLevel : int {
...
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_GE(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
-#define DCHECK_GT(val1, val2) \
-  while (false) ::arrow::util::detail::NullLog()
+#define DCHECK(condition)                     \
+  while (false) ARROW_IGNORE_EXPR(condition); \
+  while (false) ::arrow::util::Voidify() & ::arrow::util::detail::NullLog()
+#define DCHECK_OK(s)                  \
+  while (false) ARROW_IGNORE_EXPR(s); \
```
Comment (pitrou) [inline cpp/src/arrow/util/logging.h]: Previously `s` was computed (see above). This is probably the source of the segfaults on some configurations (see e.g. AppVeyor results).

Comment (pitrou) [reply cpp/src/arrow/util/logging.h]: It's confusing that `DCHECK_OK` always computes its expression but not the other `DCHECK` macros, we might remove `DCHECK_OK` and replace it with `ABORT_NOT_OK`.

Comment (bkietz) [reply cpp/src/arrow/util/logging.h]: Ah, I forgot DCHECK_OK needs to compute the expression. Thanks

**Comment by wesm** [comment]
The problem with this is `-DBUILD_WARNING_LEVEL=CHECKIN`. I do not think we ever set as a goal to have a build that passes with `-Werror` in release mode with this BUILD_WARNING_LEVEL

**Comment by wesm** [comment]
The build works fine for me locally with PRODUCTION warning level. I'm inclined to not fix this

**Comment by bkietz** [comment]
Ah, I see. In that case this can wait

**Comment by wesm** [comment]
Yes, my vote is to close as Won't Fix if others have no objections. We can add documentation that CHECKIN warning level is only meant for debug builds

**Pull request closed without merging**